src/osd/OSD.cc

```diff
@@ -6474,26 +6494,40 @@ bool OSD::ms_handle_refused(Connection *con)
   dout(2) << "ms_handle_refused con " << con << " session " << session.get() << dendl;
   if (!session)
     return false;
+
   int type = con->get_peer_type();
   // handle only OSD failures here
   if (monc && (type == CEPH_ENTITY_TYPE_OSD)) {
+    map_lock.lock_shared();
+    std::lock_guard l(mon_report_lock);
+    std::lock_guard l2(heartbeat_lock);
```

Comment:
Why do we need all three of these locks? failure_immediate_pending seems to be the only new state being accessed.

Reply:
Still not sure about these locks.

Reply:
Just a note from the RADOS meeting:

a change around locking in old & infrequently executed code makes people worried. If we want to merge this, it would be useful to have the new behavior steered by a configurable (would facilitate testing of this feature too) and bake the code in `main` for awhile. But in the first place need to understand why exactly these lock must be held here.